### DISCUSSION TITLE
Issue with Vercel API Endpoint Returning HTML Instead of JSON

### DISCUSSION THREAD

**Discussion by Hazelaway** [vercel/community#7123] [Help]
### Summary

Hi everyone,

I'm facing an issue with my web application deployed on Vercel. The application works perfectly locally but fails when deployed. Here's a summary of the problem:

- Setup: I have an Express server connected to MongoDB and a React frontend. The app is deployed on Vercel.
- 
- Problem: When making API requests (e.g., /api/current-story), the response is HTML instead of JSON. This causes a SyntaxError: Unexpected token '<', "<!doctype "... is not valid JSON error in the frontend.
- 
- What I've Tried:
* Verified that the API routes are correctly defined in vercel.json.
* Added logging to the server to ensure the endpoint is being hit.
* Created a simple test endpoint to return JSON directly.
* Checked the environment variables in Vercel.
* Reviewed multiple forum discussions and GitHub issues without a solution

Here is my vercel.json configuration:

{
  "version": 2,
  "builds": [
    {
      "src": "server/server.js",
      "use": "@vercel/node"
    },
    {
      "src": "client/package.json",
      "use": "@vercel/static-build",
      "config": {
        "distDir": "build"
      }
    }
  ],
  "routes": [
    {
      "src": "/api/(.*)",
      "dest": "/server/server.js"
    },
    {
      "src": "/(.*)",
      "dest": "/client/build/$1"
    }
  ]
}

And here is a simplified version of the problematic endpoint in server.js:
app.get('/api/current-story', (req, res) => {
  console.log('Request received at /api/current-story');
  Story.findOne({}, {}, { sort: { 'createdAt': -1 } })
    .then(story => {
      if (!story) {
        return res.status(404).json({ message: 'No story found' });
      }
      res.status(200).json({ currentPart: story.part });
    })
    .catch(err => {
      res.status(500).json({ message: 'Failed to fetch story.', error: err });
    });
});

I've tried various debugging methods, but the issue persists. Any help or insights would be greatly appreciated!

### Example

https://continuethestory.vercel.app/

### Steps to Reproduce

Hi everyone,

I'm facing an issue with my web application deployed on Vercel. The application works perfectly locally but fails when deployed. Here's a summary of the problem:

- Setup: I have an Express server connected to MongoDB and a React frontend. The app is deployed on Vercel.
- 
- Problem: When making API requests (e.g., /api/current-story), the response is HTML instead of JSON. This causes a SyntaxError: Unexpected token '<', "<!doctype "... is not valid JSON error in the frontend.
- 
- What I've Tried:
* Verified that the API routes are correctly defined in vercel.json.
* Added logging to the server to ensure the endpoint is being hit.
* Created a simple test endpoint to return JSON directly.
* Checked the environment variables in Vercel.
* Reviewed multiple forum discussions and GitHub issues without a solution

Here is my vercel.json configuration:

{
  "version": 2,
  "builds": [
    {
      "src": "server/server.js",
      "use": "@vercel/node"
    },
    {
      "src": "client/package.json",
      "use": "@vercel/static-build",
      "config": {
        "distDir": "build"
      }
    }
  ],
  "routes": [
    {
      "src": "/api/(.*)",
      "dest": "/server/server.js"
    },
    {
      "src": "/(.*)",
      "dest": "/client/build/$1"
    }
  ]
}

And here is a simplified version of the problematic endpoint in server.js:
app.get('/api/current-story', (req, res) => {
  console.log('Request received at /api/current-story');
  Story.findOne({}, {}, { sort: { 'createdAt': -1 } })
    .then(story => {
      if (!story) {
        return res.status(404).json({ message: 'No story found' });
      }
      res.status(200).json({ currentPart: story.part });
    })
    .catch(err => {
      res.status(500).json({ message: 'Failed to fetch story.', error: err });
    });
});

I've tried various debugging methods, but the issue persists. Any help or insights would be greatly appreciated!

**Comment by amyegan** [comment]
The response being HTML instead of JSON may mean that an error message is being returned instead of the expected data. You can wrap code in a [try...catch](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/try...catch) to handle the error and locate the source of the problem.

We also have a guide that covers how to adapt Express to serverless environments and how to deploy on Vercel. It may help you as well. https://vercel.com/guides/using-express-with-vercel
